File: src/main/java/com/github/sorusclient/client/adapter/v1_8_9/PotionEffectImpl.java
package com.github.sorusclient.client.adapter.v1_8_9;

import com.github.sorusclient.client.adapter.IPotionEffect;
import v1_8_9.net.minecraft.client.resource.language.I18n;
import v1_8_9.net.minecraft.entity.effect.StatusEffectInstance;

public class PotionEffectImpl implements IPotionEffect {

    protected final StatusEffectInstance effect;

    public PotionEffectImpl(StatusEffectInstance effect) {
        this.effect = effect;
    }

    @Override
    public String getDuration() {
        if (this.effect.isPermanent()) {
            return "**:**";
        }
        int duration = this.effect.getDuration();
        int seconds = duration / 20;
        int minutes = seconds / 60;
        int secondsReal = (minutes == 0 ? seconds : (seconds % (minutes * 60)));
        return minutes + ":" + (secondsReal < 10 ? "0" : "") + secondsReal;
    }

    @Override
    public String getName() {
        return I18n.translate(this.effect.getTranslationKey());
    }

    @Override
    public int getAmplifier() {
        return this.effect.getAmplifier() + 1;
    }

    @Override
    public PotionType getType() {
        switch (this.effect.getEffectId()) {
            case 1:
                return PotionType.SPEED;
            case 2:
                return PotionType.SLOWNESS;
            case 3:
                return PotionType.HASTE;
            case 4:
                return PotionType.MINING_FATIGUE;
            case 5:
                return PotionType.STRENGTH;
            case 6:
                return PotionType.INSTANT_HEALTH;
            case 7:
                return PotionType.INSTANT_DAMAGE;
            case 8:
                return PotionType.JUMP_BOOST;
            case 9:
                return PotionType.NAUSEA;
            case 10:
                return PotionType.REGENERATION;
            case 11:
                return PotionType.RESISTANCE;
            case 12:
                return PotionType.FIRE_RESISTANCE;
            case 13:
                return PotionType.WATER_BREATHING;
            case 14:
                return PotionType.INVISIBILITY;
            case 15:
                return PotionType.BLINDNESS;
            case 16:
                return PotionType.NIGHT_VISION;
            case 17:
                return PotionType.HUNGER;
            case 18:
                return PotionType.WEAKNESS;
            case 19:
                return PotionType.POISON;
            case 20:
                return PotionType.WITHER;
            case 21:
                return PotionType.HEALTH_BOOST;
            case 22:
                return PotionType.ABSORPTION;
            case 23:
                return PotionType.SATURATION;
            default:
                return PotionType.UNKNOWN;
        }
    }

}

File: src/main/kotlin/com/github/sorusclient/client/adapter/v1_18_2/PotionEffectImpl.kt
package com.github.sorusclient.client.adapter.v1_18_2

import com.github.sorusclient.client.adapter.IPotionEffect
import com.github.sorusclient.client.adapter.IPotionEffect.PotionType
import v1_18_2.net.minecraft.client.resource.language.I18n
import v1_18_2.net.minecraft.entity.effect.StatusEffect
import v1_18_2.net.minecraft.entity.effect.StatusEffectInstance

class PotionEffectImpl(protected val effect: StatusEffectInstance) : IPotionEffect {

    override val duration: String
        get() {
            if (effect.isPermanent) {
                return "**:**"
            }
            val duration = effect.duration
            val seconds = duration / 20
            val minutes = seconds / 60
            val secondsReal = if (minutes == 0) seconds else seconds % (minutes * 60)
            return minutes.toString() + ":" + (if (secondsReal < 10) "0" else "") + secondsReal
        }

    override val name: String?
        get() = I18n.translate(effect.translationKey)

    override val amplifier: Int
        get() = effect.amplifier + 1

    override val type: PotionType
        get() = when (StatusEffect.getRawId(effect.effectType)) {
            1 -> PotionType.SPEED
            2 -> PotionType.SLOWNESS
            3 -> PotionType.HASTE
            4 -> PotionType.MINING_FATIGUE
            5 -> PotionType.STRENGTH
            6 -> PotionType.INSTANT_HEALTH
            7 -> PotionType.INSTANT_DAMAGE
            8 -> PotionType.JUMP_BOOST
            9 -> PotionType.NAUSEA
            10 -> PotionType.REGENERATION
            11 -> PotionType.RESISTANCE
            12 -> PotionType.FIRE_RESISTANCE
            13 -> PotionType.WATER_BREATHING
            14 -> PotionType.INVISIBILITY
            15 -> PotionType.BLINDNESS
            16 -> PotionType.NIGHT_VISION
            17 -> PotionType.HUNGER
            18 -> PotionType.WEAKNESS
            19 -> PotionType.POISON
            20 -> PotionType.WITHER
            21 -> PotionType.HEALTH_BOOST
            22 -> PotionType.ABSORPTION
            23 -> PotionType.SATURATION
            24 -> PotionType.GLOWING
            25 -> PotionType.LEVITATION
            26 -> PotionType.LUCK
            27 -> PotionType.UNLUCK
            28 -> PotionType.SLOW_FALLING
            29 -> PotionType.CONDUIT_POWER
            30 -> PotionType.DOLPHINS_GRACE
            31 -> PotionType.BAD_OMEN
            32 -> PotionType.HERO_OF_THE_VILLAGE
            else -> PotionType.UNKNOWN
        }

}